review

    파이썬 코딩의 기본 3요소
    Data,  함수,  연결부호

Data

    <data>
    숫자   :  유일하게 컴퓨터가 인식 연산 가능한 type
    '문자' :  문자의 counting,  문자열 더하기,  slicing
    변수명  :  변수는 문자로만 가능 (내부는 임의의 값들로 수정 변환이 가능)
    tuple1, tuple2 : 여러 객체를 병렬적으로 처리 (2개 , 3개 , 여러개도 병렬처리 가능 : 꼭 짝은 맞아야 한다!!)
    ex) (tuple1, tuple2) : 컴퓨터가 연산처리 결과로 tuple을 출력시 () 소괄호로 표시
    [list형 데이터] - 숫자/ 문자
    {dict형 데이터} - { key값 : value값 } - 숫자 /문자

연결부호

    '' , ""  문자열 지정 (연산 불가능한 객체)
    . 종속   
    , 병렬나열
    : 내부설계
    = 입력
    == 대등조건
    != 부 대등조건

Function

    반복문 : for, while, loop.....   
        1.list. dict같은 반복 조건이 필요
        2. 같은 조건/ 함수/ 내용을 반복
    판단문 : if    -  True / False

## (data) string formating
formating : 문자열 내부에 변수값을 지정 및 활용

In [1]:
# %d : digit variable  
"I am %d years old in korea" %21

'I am 21 years old in korea'

In [2]:
# %s : String variable 
# "I am an %s in korea" %Account   # Account 변수에 값이 입력되어 있으면 정상실행된다.
"I am an %s in korea" %"Account"

'I am an Account in korea'

formating 에서 변수를 사용

In [3]:
# %d : digit variable  
age = 12
"I am %d years old in korea" %age

'I am 12 years old in korea'

In [4]:
# %s : String variable 
Account = 'account'
"I am an %s in korea" %Account
#"I am an %d in korea" %Account

'I am an account in korea'

In [5]:
# %s 는 다양한 입력형태를 자동으로 변환
"I am %s years old in korea" %age

'I am 12 years old in korea'

formating 변수를 2개 이상 활용

In [6]:
number = 10
day = 'three'
"I ate an %d apple. so i sick for %s days" %(number, day)

'I ate an 10 apple. so i sick for three days'

In [7]:
# %d %% : %를 화면에 출력하고 싶을 떈 double inputing을 하면 된다. 
error = 90
"this program's error is %d%%." %error

"this program's error is 90%."

    %c : character 문자 1개
    %s : String 문자열
    %d : integer 정수
    %f : floating point 부동소수

# Quiz

    "2017년 대한민국 최고 회계사" 를
    1) year = 2017 숫자를 입력한 변수을 활용하여
    2) %d 로 통해서화면에 출력해보자
    
    start>> "" %

    "2017년 대한민국 최고 회계사" 를
    1) year = 2017  숫자를 입력한 변수을 활용하여
    2) %s 로 통해서화면에 출력해보자
    
    start>> "" %

입력공간 크기값 지정

In [8]:
# default : 오른쪽 정렬
#%10  d  (전체 입력공간의 최소 크기를 지정)
"%10d" %1020

#%10  s  (전체 입력공간의 최소크기를 지정)
"%10s" %"Hello come to the real world"

#%-10  s  (전체 입력공간의 최소크기를 지정)
# -10 : 왼쪽부터 정렬
"%-10s jane" %"hello"

'      1020'

### 숫자 소숫점 자리 표현하기

In [11]:
# using float
"%0.4f"  % 3.141592

'3.1416'

In [12]:
# using float
"%.4f"  % 3.141592

'3.1416'

In [13]:
# using float
"%10.4f"  % 3.141592

'    3.1416'

# Quiz

    20000원
    2200원
    100원
    위 3줄을 모두 10칸 오른쪽 정렬로 일괄 출력해보자
    
        20000원
         2200원
          100원
    
    start>> '%10s'

# Quiz

    pi = 3.1415921213
    위 변수값을 소숫점 2자리 까지 화면에 출력해보자

### 문자열 관련 함수

In [14]:
# 문자의 갯수 세기
a = 'hobby'
a.count('ho')

1

In [15]:
# -1 : 입력값이 없으면 출력
a = 'Python is best choice'
a.find('P')

0

In [16]:
# -1 : 입력값이 없으면 출력
a = 'Life is too short'
a.index('t')

8

In [17]:
a =","
a.join('Hello World')

'H,e,l,l,o, ,W,o,r,l,d'

In [18]:
# Left strip (왼쪽의 공백 제거)
a = "    당기손이익  "
a.lstrip()

'당기손이익  '

In [19]:
# Right strip (오른쪽 공백 제거)
a = "  당기손이익   "
a.rstrip()

'  당기손이익'

In [20]:
# strip (양쪽 공백 제거)
a = "  당기손이익   "
a.strip()

'당기손이익'

In [21]:
a = "Life is too short"
a.replace("Life",  "Your leg")

'Your leg is too short'

In [22]:
# 공백을 기준으로 문자열 나눔
a = "당기손이익  (연결)"
a.split()

['당기손이익', '(연결)']

In [23]:
# 공백을 기준으로 문자열 나눔
a = "할부원금:100,000"
a.split(':')

['할부원금', '100,000']

# Quiz

    price = 20000
    .join 메소드를 활용해서  '20000원'  으로 출력해보자

# Quiz

    price = 20000원
    .replace 메소드를 활용해서  '20000krw'  으로 출력해보자

## .format 함수를 활용
index 위치값을 활용한 문자열 삽입 (포맷형식은 자동조정)

In [24]:
"I eat {0} apples".format(3)

number = 3.141592
day = "three"
"I ate {0} apples. so I was sick for {1} days".format(number, day)

# :0.2f 2자리까지 출력
number = 3.141592
day = "three"
"I ate {0:0.2f} apples. so I was sick for {1} days".format(number, day)

# :<  왼쪽 정렬
number = 3.141592
day = "three"
"I ate {0:<16} apples. so I was sick for {1} days".format(number, day)

# :>  오른쪽 정렬
number = 3.141592
day = "three"
"I ate {0:>16} apples. so I was sick for {1} days".format(number, day)

# :^  가운데 정렬
number = 3.141592
day = "three"
"I ate {0:^16} apples. so I was sick for {1} days".format(number, day)

# :-^  공백 채우기
number = 3.141592
day = "three"
"I ate {0:-^16} apples. so I was sick for {1} days".format(number, day)

'I eat 3 apples'

# Quiz

    pi = 3.14159215
    .format 함수를 활용해서
    'Pi number is 3.14'로 출력하라

    hint>> "Pi number is {0:  }".format()

# Quiz

In [31]:
from intro import naver_short_table
table = naver_short_table('035420')
rates = list(table.등락률)
rates

['-0.47%', '+0.72%', '-1.76%', '-0.93%', '-0.92%', '-2.14%']

    rates[0] 의 '-0.47%'의 '%'를 각각 지우고
    소숫점 3자리 숫자로 바꿔서 출력해보자
    
    hint>> rate[0].replace()  /100
    hint>>

In [32]:
num = float(rates[0].replace("%",""))/100

In [33]:
"%0.3f" %num

'-0.005'

for 반복문을 활용한 포매팅 변경하기

In [34]:
for rate in rates:
    num = float(rate.replace("%",""))/100
    print("%0.3f" %num)

-0.005
0.007
-0.018
-0.009
-0.009
-0.021


## [list] 리스트 자료 관리하기

#### 자료형 구분하기

In [35]:
a = [1,2,3, [11,12,13]]

In [36]:
# Slicing
a[2:]

[3, [11, 12, 13]]

In [37]:
# indexing
a[-1]

[11, 12, 13]

In [38]:
# indexing (2 dim)
a[-1][0]

11

#### Add, append,  insert data

In [39]:
b = [4,5,6]
a + b

[1, 2, 3, [11, 12, 13], 4, 5, 6]

In [40]:
b + a

[4, 5, 6, 1, 2, 3, [11, 12, 13]]

In [41]:
# 반복
a * 3

[1, 2, 3, [11, 12, 13], 1, 2, 3, [11, 12, 13], 1, 2, 3, [11, 12, 13]]

In [42]:
b.append(5)
b

[4, 5, 6, 5]

In [43]:
b.append([21,22,23])
b

[4, 5, 6, 5, [21, 22, 23]]

In [44]:
b[0] = 10
b

[10, 5, 6, 5, [21, 22, 23]]

In [45]:
# 기존값들 그대로 유지한 채 새로운 값 입력
# 해당 위치에 입력하면, 나머지는 뒤로 밀려난다
# list객체.insert(삽입할 인덱스 위치, 입력값)
b.insert(0, [31,32,33])
b

[[31, 32, 33], 10, 5, 6, 5, [21, 22, 23]]

#### Delete data

In [46]:
# pop()  insert()로 입력시 쌍을 이뤄서 실행하는 경우가 많다
b.pop()
b

[[31, 32, 33], 10, 5, 6, 5]

In [47]:
# Slicing 을 활용해서 삭제하기
b[2:5] = []
b

[[31, 32, 33], 10]

In [48]:
# index를 활용해서 삭제하기
# 빈 list가 입력되어 있다.
b[0] = []
b

[[], 10]

In [49]:
del b[0]
b

[10]

In [50]:
# value 를 기준으로 삭제하기 (처음 조건에 해당되는 값을 삭제)
b.remove(10)
b

[]

#### Sorting

In [51]:
# 기본: 오름차순 정렬
a = [5, 2, 4, 10, 1, 3]
a.sort()
a

[1, 2, 3, 4, 5, 10]

In [52]:
# 내림차순 정렬 옵션
a.sort(reverse=True)
a

[10, 5, 4, 3, 2, 1]

In [53]:
# 알파벳에도 동일한 기준으로 정렬
a = ["bike", 'apple','zoo','list','dictionary']
a.sort()
a

['apple', 'bike', 'dictionary', 'list', 'zoo']

In [54]:
# 한글에도 동일한 기준으로 정렬
a = ['일본', '중국' , '한국', '대만']
a.sort(reverse=True)
a

['한국', '중국', '일본', '대만']

#### reverse() 그대로 뒤집기

In [55]:
a = ['일본', '중국' , '한국', '대만']
a.reverse()
a

['대만', '한국', '중국', '일본']

    [ list ] 데이터 활용하기
    찾기   : [n1] indexing, [n1:n2] slicing
    더하기  : .append(객체) , .insert(index주소, 객체) , + 
    제거하기 : .remove(객체), .pop() , list[n1:n2] = [] , del list[index주소]
    정렬 : .sort(reverse=False) , .reverse()

## {dict} 딕셔너리 자료 관리하기

In [56]:
# 딕셔너리 내용 추가하기
krx_code = {'000660': 'SK하이닉스', '005390': '삼성전자', '035420': '네이버'}
krx_code['034220'] = 'LG디스플레이'
krx_code

{'000660': 'SK하이닉스', '005390': '삼성전자', '034220': 'LG디스플레이', '035420': '네이버'}

In [57]:
del krx_code['000660']
krx_code

{'005390': '삼성전자', '034220': 'LG디스플레이', '035420': '네이버'}

In [58]:
krx_code2 = {'000661': 'SK하이닉스', '005391': '삼성전자', '035421': '네이버'}

In [59]:
# krx_code + krx_code2
for code in krx_code2:
    if code not in krx_code:
        krx_code[code] = krx_code2[code]

In [60]:
krx_code

{'000661': 'SK하이닉스',
 '005390': '삼성전자',
 '005391': '삼성전자',
 '034220': 'LG디스플레이',
 '035420': '네이버',
 '035421': '네이버'}

    {dict}의 자료 관리
    기본적 Key 값들만 출력된다.
    {dict}[key값] 을 해야 value 를 입력 / 출력

    제거 : key 값 만으로 삭제가능
    추가 : dict객체이름[ key값 ] = value값
    연결 : for 반복문을 통해서 연결한다.

# Quiz

    krx_code 자료를
    krx_code.items()를 활용해서
    code 와 '회사이름' 의 key , value 값을 
    '회사이름' 과 code 의 key , value 형태로 변환해서 출력해라
    
    hint>>  code = {} 
            for  튜플자료형   in   krx_code.items():
                    code['key값'] = 'value 값'

In [61]:
code = {}
for k , v in krx_code.items():
    code[v] = k
code

{'LG디스플레이': '034220', 'SK하이닉스': '000661', '네이버': '035420', '삼성전자': '005390'}

### Boolean 함수
True, False : Masking 역활로 활용

In [62]:
a = [1,2,3,4,5,6,7,8,9]

In [63]:
10 in a

False

In [64]:
6 in a

True

In [65]:
10 not in a

True

In [66]:
'a' in 'Account'

False

In [67]:
'a' in 'account'

True

#### if

In [68]:
pocket = ['paper','cellphone']

In [69]:
if 'paper'  in pocket:
    print('print the paper')
elif 'money' in pocket:
    print('have a money')
else:
    print("have nothing")

print the paper


In [70]:
if 'paper'   in pocket: print('print the paper')
elif 'money' in pocket: print('have a money')
else:                   print("have nothing")

print the paper


### [ list 내부에서 활용하는  함수 ]

In [71]:
a = [1,2,3,4,5,6,7,8,9]

In [72]:
# list 자료를 활용
for num in a:
    if num > 5:
        print(num)

6
7
8
9


In [73]:
data = [num     for num in a     if num>5]
data

[6, 7, 8, 9]

In [74]:
# dict 자료를 활용
for code in krx_code:
    if code != '005390':
        print(krx_code[code])

네이버
삼성전자
네이버
SK하이닉스
LG디스플레이


In [75]:
code =  [ krx_code[code]  for code in krx_code   if code != '005391']
code

['네이버', '네이버', 'SK하이닉스', '삼성전자', 'LG디스플레이']

### for 함수
반복문 with  Range

range(start, end, step)

In [76]:
for i in range(10): 
    print(i)

0
1
2
3
4
5
6
7
8
9


In [77]:
[ i   for i  in range(10)]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [78]:
[ i   for i  in range(1,10)]

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [79]:
[ i   for i  in range(2,10,2)]

[2, 4, 6, 8]

In [80]:
[ i   for i  in range(2,11,2)]

[2, 4, 6, 8, 10]

slicing [ start : end : step ] 

In [81]:
data = [ i  for i in range(0,100)]

In [82]:
data[1]

1

In [83]:
data[1:10]

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [84]:
data[::10]

[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]

# Quiz

    [list 내부활용 for 문]을 사용해서
    data = 0 ~ 200까지 숫자 list를 생성 후
    data의 20 단위의 숫차를 출력해라
    
    hint>>  range( start, end, step )
            slicing[ start : end : step ]